extmod/modbluetooth: Addressing modes, Unix H4 UART, reliability fixes

jimmo opened the pull request:
Work done in collaboration with @andrewleech

This started as just an implementation of NimBLE on Unix but ended up with lots of dependent changes and bugs fixed along the way. Ideally this would have been multiple PRs but I've tried to split each part into distinct commits. Other than the binding refactor (which is unfortunately very difficult to split), each commit is quite small.

Bug fixes:
* Fix handling of duration_ms on gap_scan
* Fix UART RX IDLE handler, improves latency on handling incoming HCI UART data.
* Implement the NimBLE mutex
* Fix race between the READ_REQUEST "hard" IRQ and the other scheduler-invoked IRQs
* Made it possible to use active scan mode.
* Writing to a characteristic (as server) now notifies subscribed clients.
* STM32 HCI UART timeout was too low (leading to occasional TX failures leading to loss of HCI sync)

Code cleanup:
* Refactoring to simplify port/driver/stack/binding interfaces, improving the startup and shutdown process.
* Make debugging/logging macros consistent and clearer across the BLE-related files

Unix port:
* Allow nimble_malloc to work on Unix
* Implement support for H4 UART (typically via USB UART adaptor), via btstack and NimBLE
  This allows (for example) a Zephyr HCI controller running on an nRF to be used from the Unix port.
* Extract out common code from the libusb mode and the stm32 port.

Addressing:
* Clarify behavior for obtaining an identity address (use controller's public address and then fallback to static address)
* On btstack, allow using a Zephyr controller's static address (generated from the FICR register on nRF)
* On btstack+nimble, allow using mp_hal_get_mac as as a static address.
* Implement a Python API for switching between public and static address modes (and preliminary support for RPA/NRPA)
* Make querying the MAC address return the currently used address.
* Implement preliminary support for MAC rotation (when using RPA/NRPA)

This is tested using the multitests on:
* PYBD btstack
* PYBD nimble
* Unix USB
* Unix btstack
* Unix nimble
* ESP32 nimble

More testing required on
* WB55 bstack
* WB55 nimble
(I'm working on those right now)

TODO:
* Notify subscribed clients for btstack.
* Add multitests for address modes and subscriptions.

pi-anl committed:
extmod/nimble: Automatically notify subscribed connections on gatts_write().

pi-anl committed:
extmod/bluetooth: Support active scan in BLE.gap_scan().

jimmo committed:
tests/multi_bluetooth: Update to new config('mac') behaviour.

jimmo committed:
docs/library/ubluetooth.rst: Add address modes.

jimmo committed:
ports/stm32/machine_uart.c: Allow static IRQ handler registration.

This will allow the HCI UART to use a non-heap mp_irq_obj_t, which avoids needing to make a root pointer for it.

Misc cleanups to mpirq.c to remove unused functionality.

jimmo committed:
bluetooth: Refactor stack/hci/driver/port bindings.

Previously the interaction between the different parts was different on each port and each stack. This commit defines common interfaces between them and implements them for cyw43, btstack, nimble, stm32, unix.

jimmo committed:
extmod/btstack: Implement scan duration_ms.

jimmo committed:
ports/unix: Implement BLE H4 HCI UART for btstack/nimble.

jimmo committed:
extmod/modbluetooth: Implement configuration of address modes.

jimmo committed:
extmod/modbluetooth: Rename logging macro.

Also use correct format specifier for %zu.

Always prefix with "btstack:" or "nimble:".

jimmo committed:
extmod/nimble: Make nimble_malloc work with allocated size.

jimmo committed:
extmod/nimble: Implement NimBLE mutex.

jimmo committed:
stm32/modbluetooth_hci.c: Use a static mp_irq_obj_t for HCI UART.

jimmo committed:
ports/unix: Always enable -f*-sections.

pi-anl committed:
extmod/bluetooth: Add btstack support for _IRQ_GATTS_READ_REQUEST.

pi-anl committed:
ports/stm32/mphciport.c: Increase UART timeout.

2ms used previously was not long enough and the result is that it loses HCI
sync.

Also print error on tx failure to make this more obvious in the future.

jimmo committed:
extmod/modbluetooth: Fix race between READ_REQUEST and other IRQs.

jimmo committed:
extmod/modbluetooth: Allow using mp_hal_get_mac as a static address.

Generally a controller should either have its own public address hardcoded, or
loaded by the driver (e.g. cywbt43).

However, for a controller that has no public address where you still want a
long-term stable address, this allows you to use a static address generated by
the port. Typically on STM32 this will be an LAA, but a board might override
this.

jimmo committed:
extmod/nimble: Set struct alignment correctly on 64-bit arch.

andrewleech commented on ports/unix/mphciport.c:
```diff
@@ -0,0 +1,198 @@
+    int flags = O_RDWR | O_NOCTTY | O_NONBLOCK;
+    uart_fd = open(uart_device_name, flags);
+    if (uart_fd == -1) {
+        DEBUG_printf("Unable to open port %s", uart_device_name);
```
This would be better as an exception that's then user visible - currently if the port is missing or cannot be opened it's a silent fail here and the rest of the startup just looks broken

jimmo replied:
Done -- and added the plumbing to make this work for btstack and nimble. On btstack it can detect the init failure, on nimble it now leads to an init timeout.

andrewleech commented on ports/stm32/mphciport.c:
```diff
@@ -135,60 +123,83 @@ int mp_bluetooth_hci_uart_readchar(void) {
+    uart_tx_data(&mp_bluetooth_hci_uart_obj, (void *)buf, len, &errcode);
+    if (errcode != 0) {
+        printf("\nmp_bluetooth_hci_uart_write: failed to write to UART %d\n", errcode);
+    }
```
In my branch I've changed this if to `assert(errcode == 0);`

My updated commit is here, it includes your comment and removes the extra comment/printf below:
https://github.com/andrewleech/micropython/commit/32bf8073d9dc843abc95aca48ed369ab854db8f6

jimmo replied:
I think this is too important to be a debug-only notification...

andrewleech replied:
Ah yeah, I forgot that asserts are disabled in non-debug builds... yep perhaps a static flag that can be checked at a higher layer will be needed, perhaps something that can then trigger a stack/hci reset.

jimmo committed:
extmod/btstack: Detect HCI UART init failure.

jimmo committed:
ports/esp32: Pin MP and NimBLE to core 0.

MP and NimBLE must be on the same core (for synchronisation of the BLE ringbuf and the MP scheduler).

However, in the current IDF versions (3.3 and 4.0) there are issues (e.g. #6343) with running NimBLE on core 1.

This change makes it possible to reliably run the BLE multitests.

jimmo committed:
extmod/nimble: Add timeout for HCI sync on startup.

This allows `ble.active(1)` to fail correctly if the HCI controller is
unavailable.

Also avoids an infine loop in the NimBLE event handler where NimBLE doesn't
correctly detect that the HCI controller is unavailable and keeps trying to
reset.

Also fixes an issue where GATT service registrations were left allocated,
which led to a bad realloc if the stack was activated multiple times.

dpgeorge commented:
The following commits from this PR have been applied to master (they are minor and fix existing bugs) (the hash refers to the commit in master):
- 3d9a7ed02f4e50ee2eb541ac2f4d7f6f5b5d316e extmod/btstack: Implement GAP scan duration_ms parameter.
- 0bc2c1c1057d7f5c1e4987139062386a8f9fe5f2 extmod/modbluetooth: Fix race between READ_REQUEST and other IRQs.
- a80a146858b3c9a7fbae4030a524666ab19f7a47 extmod/bluetooth: Support active scanning in BLE.gap_scan().

dpgeorge commented:
Superseded by #6405 (with some notes there about what was not included).

dpgeorge commented:
Changes here were merged (somewhat modified) through #6405, see 23109988c29447067b15f1943f72c0b486819c6c through 126f972c34f25bffc07d0d1cb3e536921fec0979

dpgeorge commented:
Thanks @jimmo and @andrewleech for all the effort!
